### ISSUE TITLE
keras.backend.function() broken in 2.2.0 for string inputs with TF backend

### ISSUE THREAD

**Issue by ageron** [keras-team/keras#10459]
The following code works on Keras 2.1.6 using TensorFlow 1.9.0-rc0, but it fails when upgrading to Keras 2.2.0:

```python
import tensorflow as tf
import numpy as np
from keras.backend import function

x = tf.placeholder(shape=[None], dtype=tf.string)
f = tf.substr(x, 2, 3)
func = function(inputs=[x], outputs=[f])

some_strings = np.array([b"abcdefg", b"012345689"])
print(func([some_strings]))
```

Excepted output is:

```
[array([b'cde', b'234'], dtype=object)]
```

But here is the exception I get instead with Keras 2.2.0:

```pycon
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ageron/.virtualenvs/ml/lib/python3.5/site-packages/keras/backend/tensorflow_backend.py", line 2661, in __call__
    return self._call(inputs)
  File "/home/ageron/.virtualenvs/ml/lib/python3.5/site-packages/keras/backend/tensorflow_backend.py", line 2614, in _call
    dtype=tensor.dtype.base_dtype.name))
  File "/home/ageron/.virtualenvs/ml/lib/python3.5/site-packages/numpy/core/numeric.py", line 492, in asarray
    return array(a, dtype, copy=False, order=order)
TypeError: data type "string" not understood
```

Running on Ubuntu 16.4, Python 3.5.2, and using the following library versions:
```pycon
>>> import tensorflow as tf
>>> tf.__version__
'1.9.0-rc0'
>>> import numpy as np
>>> np.__version__
'1.14.3'
>>> import keras
Using TensorFlow backend.
>>> keras.__version__
'2.2.0'
```

This issue may be related to #10372.

**Comment by ageron** [comment]
The problem comes from this call:

```python
np.asarray(value, dtype=tensor.dtype.base_dtype.name)
```

This is because `tensor.dtype.base_dtype.name` is equal to `"string"`, and NumPy does not support that type. Perhaps a simple fix could be to map `"string"` to `"object"`?

**Comment by ageron** [comment]
I submitted a quick fix in PR #10460 , simply mapping `tensor.dtype.base_dtype.name` to `object` when it evaluates to `string`.
